internal/endpoint/pop3/pop3.go

```diff
@@ -0,0 +1,466 @@
+*/
+
+package pop3
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+	"strconv"
+
+	"github.com/kiwiz/popgun"
+	pop3backend "github.com/kiwiz/popgun/backends"
+	"github.com/emersion/go-imap"
+	sortthread "github.com/emersion/go-imap-sortthread"
+	imapbackend "github.com/emersion/go-imap/backend"
+	_ "github.com/emersion/go-message/charset"
+	i18nlevel "github.com/foxcpp/go-imap-i18nlevel"
+	"github.com/foxcpp/go-imap-mess"
+	"github.com/foxcpp/maddy/framework/config"
+	modconfig "github.com/foxcpp/maddy/framework/config/module"
+	tls2 "github.com/foxcpp/maddy/framework/config/tls"
+	"github.com/foxcpp/maddy/framework/log"
+	"github.com/foxcpp/maddy/framework/module"
+	"github.com/foxcpp/maddy/internal/auth"
+	"github.com/foxcpp/maddy/internal/authz"
+	"github.com/foxcpp/maddy/internal/proxy_protocol"
+	"github.com/foxcpp/maddy/internal/updatepipe"
+)
+
+type Endpoint struct {
+	addrs         []string
+	serv          *popgun.Server
+	listeners     []net.Listener
+	proxyProtocol *proxy_protocol.ProxyProtocol
+	Store         module.Storage
+
+	tlsConfig   *tls.Config
+	listenersWg sync.WaitGroup
+
+	saslAuth auth.SASLAuth
+
+	storageNormalize authz.NormalizeFunc
+	storageMap       module.Table
+
+	Log log.Logger
+}
+
+func New(modName string, addrs []string) (module.Module, error) {
+	endp := &Endpoint{
+		addrs: addrs,
+		Log:   log.Logger{Name: modName},
+		saslAuth: auth.SASLAuth{
+			Log: log.Logger{Name: modName + "/sasl"},
+		},
+	}
+
+	return endp, nil
+}
+
+func (endp *Endpoint) Init(cfg *config.Map) error {
+	var (
+		insecureAuth bool
+		debug        bool
+		errors       bool
+	)
+
+	cfg.Callback("auth", func(m *config.Map, node config.Node) error {
+		return endp.saslAuth.AddProvider(m, node)
+	})
+	cfg.Bool("sasl_login", false, false, &endp.saslAuth.EnableLogin)
+	cfg.Custom("storage", false, true, nil, modconfig.StorageDirective, &endp.Store)
+	cfg.Custom("tls", true, true, nil, tls2.TLSDirective, &endp.tlsConfig)
+	cfg.Custom("proxy_protocol", false, false, nil, proxy_protocol.ProxyProtocolDirective, &endp.proxyProtocol)
+	cfg.Bool("insecure_auth", false, false, &insecureAuth)
+	cfg.Bool("errors", false, false, &errors)
+	cfg.Bool("debug", true, false, &debug)
+	config.EnumMapped(cfg, "storage_map_normalize", false, false, authz.NormalizeFuncs, authz.NormalizeAuto,
+		&endp.storageNormalize)
+	modconfig.Table(cfg, "storage_map", false, false, nil, &endp.storageMap)
+	config.EnumMapped(cfg, "auth_map_normalize", true, false, authz.NormalizeFuncs, authz.NormalizeAuto,
+		&endp.saslAuth.AuthNormalize)
+	modconfig.Table(cfg, "auth_map", true, false, nil, &endp.saslAuth.AuthMap)
+	if _, err := cfg.Process(); err != nil {
+		return err
+	}
+
+	if updBe, ok := endp.Store.(updatepipe.Backend); ok {
+		if err := updBe.EnableUpdatePipe(updatepipe.ModeReplicate); err != nil {
+			endp.Log.Error("failed to initialize updates pipe", err)
+		}
+	}
```

Comment:
The `updatepipe` stuff is entirely unused. I'm not sure if there's any important side effects or if it can be removed.

Reply:
It is used for IMAP update synchronization when multiple servers are using the same DB, can be safely removed for POP3.